spec: Rust like pseudo code for Sequential supervisor and core verification (#509)
informalsystems/tendermint-rs: docs/spec/supervisor/supervisor.md

```rust

> /// Stores the Light Blocks which could not be confirmed to have originated from

Reviewer: ```suggestion
/// Stores the LightBlocks which could not be confirmed to have originated from
```

/// the same chain as the predecesor light blocks.

> type FetchedStack = Stack<LightBlock>;

Reviewer: What are the semantics of the stack? I don't believe there is a type definition in this file.

Author: There is no semantics as we could use a `Vec` for this, the goal was just to name it here and explain it is going to be used as a stack (push, pop, peek, is_empty) and not with random access.

> /// Pair of light block and height of the light block which was highest 

Reviewer: ```suggestion
/// Pair of light block and height of the light block which was the 
```

/// highest one for which we believe originated from the same chain

> type ChainPair = (LightBlock, Height);

Reviewer: The naming is confusing as it indicates that a `Chain` is paired with something. What I'm also wondering doesn't the `LightBlock` contain its `Height`, what's the benefit of pairing it with the `Height`.

Author: Height in this case is the height of the block which was used to check the light. For example if we inserted a light block `A` into the `Trace` after passing all the checks in verify_to_target using light block B of height `h`. The `ChainPair` for `A` would be (`A`, `h`).

Reviewer: Is it always the height of `B` -1? In any case the term `Height` is so ubiquitous everywhere that we should encode what height we are referring to here, which in turn an help challenge how it can be constructed.

Reviewer: Still the term `ChainPair` for a pairing of block with height is miss-leading, imho.

Author: Made a custom `enum VerificationChain`.

/// Traces the history of verify_to_target.
trait Trace {
    /// Returns the height of the current highest inserted

>     fn current_height(&self) -> Height;

Reviewer: What's the use-case for just getting the current height of the `Trace`, can't you obtain that through the `current` `LightBlock`?

Author: Just as a convenience method, can be removed.

    /// Returns the current highest inserted light block.
    fn current(&self) -> &LightBlock;

    /// Insert a new highest light block. assert(self.current_height() < light_block.height())
    fn insert(&mut self, light_block: LightBlock) -> Result<(), Error>;

    /// Returns the list of all LightBlock which were inserted through 
    /// insert with associated previous current_height().

>     fn verification_chain(&self) -> LinkedList<ChainPair>;

Reviewer: ```suggestion
    fn chain(&self) -> LinkedList<ChainPair>;
```

This seems expressive enough.

}

/// Given a primary and a starting_light_block confirm that the light block of target_height received
/// from primary for target_height is from the same chain as the starting_light_block.
/// Solution for forward skipping verification.
pub fn verify_to_target(
    primary: PeerID,
    starting_light_block: LightBlock,

>     target_height: Height) -> Result<Trace, (LightBlock, Reason)> {

Reviewer: The `Reason` type is not expanded anywhere. Would be helpful to see what it actually looks like. It's uncommon to not

Author: It's a mistake it should be [`Error`](https://github.com/informalsystems/tendermint-rs/blob/548d76d8b7025f1a3ad1ee9adf32a5e6d24e2197/light-client/src/errors.rs#L18).

    // Verified state always has one starting LightBlock

>     let mut trace = Trace::from(startingLightBlock);

Reviewer: ```suggestion
    let mut trace = Trace::from(starting_light_block);
```

    let mut fetch_stack = FetchedStack::empty();

    // insert the light block of target height, simplifies the loop
    // we can say that at each iteration there should be something on stack

>     fetch_stack.push(fetch_light_block(primary, target_height));

Reviewer: What does `fetch_light_block` do?

Author: It's just copied from the current specifications. In current code it's from the [`Io`](https://github.com/informalsystems/tendermint-rs/blob/5e298723f6f60ccf59b60c5bc7c202b338238260/light-client/src/components/io.rs#L71) component.

Reviewer: It would greatly benefit if this side-effect is not done in this function, it could be a parameter to be called with.

Author: This was an error. The structure is changed a but now.

        let lowest_unverified = fetch_stack.peek().unwrap().clone();
        
        // do the necessary checks
        let verdict = valid_and_verified(trace.current(), &lowest_unverified);

        match verdict {
            // validated LightBlock, improve the verified_state
            Valid => {
                fetch_stack.pop();
                trace.insert(lowest_unverified);
                continue;
            },
            // invalid block stop
            Invalid(reason) => {
                return Err((lowest_unverified, reason))
            },
            Untrusted => {
                // we need an intermediate LightBlock
                current_height = 
                    compute_height(trace.current_height(), lowest_unverified.height());
            }
        }
    }

    assert_eq!(trace.current_height(), target_height);
    assert!(fetch_stack.is_empty());

    return Ok(trace);
}

/// LightStore no longer has notions of Trusted, Verified, Unverified and Failed.
/// The LightBlocks which are stored are all assumed to have originated from the
/// same chain, irrespective of the peer from which we received it. The store

> /// provides the invariant that there is a 1-1 mapping from height to a LightBlock.

Reviewer: If we want to ensure the 1-1 invariant we should think in terms of a `Set` where we have a uniqueness constraint on the height, unless I miss-understand this part.

Author: I think it would make more sense to have a `Map<Height, LightBlock>` where keys would be automatically unique.

Reviewer: For clarification, every height can only exist once? And there is only exactly one `LightBlock` per unique height?

Author: Yes, that was the intention. At each point in time, there can be only one `LightBlock` which can be considered to be from the valid blockchain. Conflict on a single height would mean there is a fork.

trait LightStore {
    /// Current highest LightBlock which we believe to be from a chain 
    /// for which this store is associated with.
    fn highest(&self) -> LightBlock;

    /// Stores the list of ChainPair, possibly changing the highest LightBlock;
    /// Precondition is the existence of Light Block for which the chain can be linked

>     /// with the already stored ChainPair. Failing if the chain can bot be linked with

Reviewer: ```suggestion
    /// with the already stored ChainPair. Failing if the chain can not be linked with
```

    /// existing history or if there is conflicting information.
    fn store_chain(&mut self, chain: LinkedList<ChainPair>) -> Result<(), Error>;
    
    /// Stores one ChainLink, failing if there is conflicting information in the store.
    /// (forks, etc.). Can be used when light blocks are received out of order.
    fn store_pair(&mut self, link: ChainLink) -> Result<(), Error>;

    /// Get verification chain of LightBlock ending at height high and starting starting from low.

>     fn get_chain(&self, low: Height, high: Height) -> LinkedList<ChainPair>;

Reviewer: This operation can fail and should likely return a `Result`.

    /// In a case a fork is detected allows the light store to recover for a specified boundary
    /// height. Erases all light blocks from the store with height greater or equal to boundary.
    /// Returning the chain of light blocks which are removed. To be used if a fork is discovered
    /// after insertions of light blocks.
    fn recover(&mut self, boundary: Height) -> LinkedList<ChainPair>;

    /// Returns the firsts LightBlock above if it exists.
    fn above(&self, height: Height) -> Option<&LightBlock>;
    
    /// Returns the firsts LightBlock below if it exists.
    fn below(&self, height: Height) -> Option<&LightBlock>;
    
    /// Tries to get a certain height.

>     fn get(&self, height: Height) -> Option<&LightBlock>;

Reviewer: I'd model all of these as methods that return `Result`s.

    /// with the already stored VerificationChain. Failing if the chain can not be linked with
    /// existing history or if there is conflicting information.
    fn store_chain(&mut self, chain: VerificationChain) -> Result<(), Error>;
    
    /// Stores one VerificationChain, failing if there is conflicting information in the store.
    /// (forks, etc.). Can be used when light blocks are received out of order.
    fn store_pair(&mut self, link: VerificationChain) -> Result<(), Error>;

    /// Get verification chain of LightBlock ending at height high and starting starting from low.
    fn get_chain(&self, low: Height, high: Height) -> VerificationChain;

    /// In a case a fork is detected allows the light store to recover for a specified boundary
    /// height. Erases all light blocks from the store with height greater or equal to boundary.
    /// Returning the chain of light blocks which are removed. To be used if a fork is discovered
    /// after insertions of light blocks.
    fn recover(&mut self, boundary: Height) -> Result<VerificationChain, LightStoreError>;

    /// Returns the firsts LightBlock above if it exists.

>     fn above(&self, height: Height) -> Result<&LightBlock, ListStoreError>;

Reviewer 2: why do you need `above` and `below` when you already have `get`?

    
    /// Returns the firsts LightBlock below if it exists.
    fn below(&self, height: Height) -> Result<&LightBlock, ListStoreError>;
    
    /// Tries to get a certain height.
    fn get(&self, height: Height) -> Result<&LightBlock, ListStoreError>;
}

/// Current understanding how a LightNode functions.
pub fn sequential_supervisor() -> Result<(), Forked> {
    loop {
	    // get the next height
        let target_height = input();
        // light_store contains all (LightBlock, previous), for this supervisor all have passed
        // fork detection and there will be no recoveries, etc.
		
		// Verify
        let mut result = Err(NoPeer);

>         // try with primaries until you succeed or there are no more peers

Reviewer 2: don't you have a single primary at any moment?

>         while result.is_err() {

Reviewer 2: looks like this will never terminate, making the rest of the code unreachable
